core/src/main/java/org/elasticsearch/index/shard/ShardSplittingQuery.java

```diff
@@ -0,0 +1,245 @@
+import org.apache.lucene.index.TermsEnum;
+import org.apache.lucene.search.ConstantScoreScorer;
+import org.apache.lucene.search.ConstantScoreWeight;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.TwoPhaseIterator;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.BitSetIterator;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.FixedBitSet;
+import org.elasticsearch.Version;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.cluster.routing.OperationRouting;
+import org.elasticsearch.index.mapper.IdFieldMapper;
+import org.elasticsearch.index.mapper.RoutingFieldMapper;
+import org.elasticsearch.index.mapper.Uid;
+
+import java.io.IOException;
+import java.util.function.IntConsumer;
+import java.util.function.Predicate;
+
+/**
+ * A query that selects all docs that do NOT belong in the current shards this query is executed on.
+ * It can be used to split a shard into N shards marking every document that doesn't belong into the shard
+ * as deleted. See {@link org.apache.lucene.index.IndexWriter#deleteDocuments(Query...)}
+ */
+final class ShardSplittingQuery extends Query {
+    private final IndexMetaData indexMetaData;
+    private final int shardId;
+
+    ShardSplittingQuery(IndexMetaData indexMetaData, int shardId) {
+        if (indexMetaData.getCreationVersion().before(Version.V_6_0_0_rc2)) {
+            throw new IllegalArgumentException("Splitting query can only be executed on an index created with version "
+                + Version.V_6_0_0_rc2 + " or higher");
+        }
+        this.indexMetaData = indexMetaData;
+        this.shardId = shardId;
+    }
+
+    @Override
+    public Weight createWeight(IndexSearcher searcher, boolean needsScores, float boost) {
+        return new ConstantScoreWeight(this, boost) {
+            @Override
+            public String toString() {
+                return "weight(delete docs query)";
+            }
+
+            @Override
+            public Scorer scorer(LeafReaderContext context) throws IOException {
+                LeafReader leafReader = context.reader();
+                FixedBitSet bitSet = new FixedBitSet(leafReader.maxDoc());
+                Terms terms = leafReader.terms(RoutingFieldMapper.NAME);
+                Predicate<BytesRef> includeInShard = ref -> {
+                    int targetShardId = OperationRouting.generateShardId(indexMetaData,
+                        Uid.decodeId(ref.bytes, ref.offset, ref.length), null);
+                    return shardId == targetShardId;
+                };
+                if (terms == null) { // this is the common case - no partitioning and no _routing values
+                    findSplitDocs(IdFieldMapper.NAME, includeInShard, leafReader, bitSet::set);
```

Comment:
maybe an `assert  indexMetaData.isRoutingPartitionedIndex() == false` would help readers realize that we require routing values for partitioned indices